transport/socks5/socks5.go

```diff
@@ -181,23 +190,24 @@ func ClientHandshake(rw io.ReadWriter, addr Addr, command Command, user *User) (
 		return nil, errors.New("socks version mismatched")
 	}
 
-	if buf[1] == 0x02 /* USERNAME/PASSWORD */ {
+	if buf[1] == AuthUsernamePassword /* USERNAME/PASSWORD */ {
 		if user == nil {
 			return nil, errors.New("auth required")
 		}
 
+		authMsgLen := 1 + 1 + len(user.Username) + 1 + len(user.Password)
+		if authMsgLen > MaxAuthLen {
+			return nil, errors.New("auth message too long")
+		}
+
 		// password protocol version
-		authMsg := &bytes.Buffer{}
```

Comment:
`bytes.Buffer` has `Grow` method, so we can do like:

```go
authMsg := &bytes.Buffer{}
authMsg.Grow(authMsgLen)
```

Reply:
Grow is not the one to go, because for every Grow function, it allocates more capacity than you actually need, if you already know the exact of the data you are going to store,

```
username := "fdasfksda"
password := "a very very very very very very very very very very very long password"

...

authMsg = &bytes.Buffer{}
authMsg.Grow(authMsgLen)

//after
//cap: 82 len: 0
//cap: 82 len: 1
//cap: 82 len: 2
//cap: 82 len: 11
//cap: 82 len: 12
//cap: 82 len: 82
//Grow cap: 96 len: 0
```

Reply:
it behaves like Java arraylist where it always give you some head room, which is really needed in this case.